jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,335 @@
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
+            if(graph.getEdgeWeight(e) < 0) {
+                // initialize all paths of cost < 0
+                LinkedList<V> pathVertices = new LinkedList<>();
+                Map<V, Integer> pathLabels = new LinkedHashMap<>();
+                V sourceVertex = graph.getEdgeSource(e);
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.put(sourceVertex, labels.get(sourceVertex));
+                pathLabels.put(targetVertex, labels.get(targetVertex));
+                LabeledPath<V> path = new LabeledPath<>(pathVertices, graph.getEdgeWeight(e), pathLabels);
+
+                // add path to set of paths of length 1
+                pathsLengthK.add(path);
+            }
+        }
+
+        while(k < lengthBound && cycleResult.getCost() >= 0) {
+            while(!pathsLengthK.isEmpty()) {
+
+                // go through all valid paths of length k
+                for(Iterator<LabeledPath<V>> it = pathsLengthK.iterator(); it.hasNext();) {
+                    LabeledPath<V> path = it.next();
+                    it.remove();
+
+                    V head = path.getHead();
+                    V tail = path.getTail();
+
+                    if(graph.containsEdge(tail, head) && path.getCost() + graph.getEdgeWeight(graph.getEdge(tail, head)) < cycleResult.getCost()) { // the path builds a valid cycle
+                        cycleResult = path.clone();
+                        cycleResult.addVertex(head, graph.getEdgeWeight(graph.getEdge(tail, head)), labels.get(head));
+
+                        /*
+                         * only return the cycle if it is negative (in the first iteration it can be non-negative)
+                         */
+                        if(cycleResult.getCost() < 0) {
+                            return cycleResult;
+                        }
+                    }
+
+                    for(E e : graph.outgoingEdgesOf(tail)) {
+                        V currentVertex = graph.getEdgeTarget(e);
+                        if(!path.getLabels().contains(labels.get(currentVertex)) && path.getCost() + graph.getEdgeWeight(e) < 0) { // extend the path if the extension is still negative a correctly labeled
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, graph.getEdgeWeight(e), labels.get(currentVertex));
+                            pathsLengthKplus1.add(newPath);
+
+                            /*
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            testDomination(path, pathsLengthKplus1);
```

Comment:
Shouldn't you be testing newPath rather than path?  If that's true, then it's probably simpler to defer adding it until after testDomination.